### PR TITLE
Fix metrics reported as written but not actually written 

### LINKED ISSUE #9514: outputs.influxdb not buffering points on telegraf 1.19.1
With telegraf 1.19.1, internal_write reports values of buffer_size for outputs.influxdb that are smaller than metric_batch_size even when the influxdb instance is down and more points have been generated. Here is a configuration for comparison also with http output which buffers points correctly (as with previous versions).

### Relevant telegraf.conf:
```toml
[global_tags]
  region = "eu-west-1"

[agent]
  metric_buffer_limit = 100000
  flush_interval = "2s"

[[inputs.internal]]
  interval="4ms"

[[outputs.influxdb]]
  urls = [ "http://localhost:8086" ]
  skip_database_creation = true

[[outputs.http]]
  url = "http://127.0.0.1:8080/telegraf"

[[outputs.file]]
  files = [ "telegraf.out" ]
```

### System info:
Linux on AMD64 with telegraf 1.19.1 (https://dl.influxdata.com/telegraf/releases/telegraf-1.19.1_linux_amd64.tar.gz) and 1.18.3 (https://dl.influxdata.com/telegraf/releases/telegraf-1.18.3_linux_amd64.tar.gz)
No processes are listening on localhost:8086 nor 127.0.0.1:8080.

### Steps to reproduce:
1. Use the above configuration with 1.19.1 binary and run it for 20s. Rename telegraf.out to telegraf-1.19.1.out
2. Then run the same configuration with 1.18.3 version binary for 20s, and rename telegraf.out to telegraf-1.18.3.out
3. Compare the last lines of telegraf-1.19.1.out and telegraf-1.18.3.out
$ tail telegraf-1.19.1.out | grep write | tail -3
internal_write,host=xxxxx,output=influxdb,region=eu-west-1,version=1.19.1 metrics_written=29000i,metrics_dropped=0i,buffer_size=682i,buffer_limit=100000i,metrics_filtered=0i,write_time_ns=1823419i,errors=29i,metrics_added=29682i 1626627449853000000
internal_write,host=xxxxx,output=http,region=eu-west-1,version=1.19.1 metrics_dropped=0i,buffer_size=29682i,buffer_limit=100000i,metrics_filtered=0i,write_time_ns=11507427i,errors=0i,metrics_added=29682i,metrics_written=0i 1626627449853000000
internal_write,host=xxxxx,output=file,region=eu-west-1,version=1.19.1 metrics_dropped=0i,buffer_size=682i,buffer_limit=100000i,metrics_filtered=0i,write_time_ns=17995062i,errors=0i,metrics_added=29682i,metrics_written=29000i 1626627449853000000
$ tail telegraf-1.18.3.out | grep write | tail -3
internal_write,host=xxxxx,output=influxdb,region=eu-west-1,version=1.18.3 metrics_dropped=0i,buffer_size=29718i,buffer_limit=100000i,metrics_filtered=0i,write_time_ns=3341971i,errors=29i,metrics_added=29718i,metrics_written=0i 1626627746465000000
internal_write,host=xxxxx,output=http,region=eu-west-1,version=1.18.3 buffer_limit=100000i,metrics_filtered=0i,write_time_ns=13574467i,errors=0i,metrics_added=29718i,metrics_written=0i,metrics_dropped=0i,buffer_size=29718i 1626627746465000000
internal_write,host=xxxxx,output=file,region=eu-west-1,version=1.18.3 metrics_added=29718i,metrics_written=29000i,metrics_dropped=0i,buffer_size=718i,buffer_limit=100000i,metrics_filtered=0i,write_time_ns=20608019i,errors=0i 1626627746465000000

### Expected behavior:
internal_write should report similar numbers for output=influxdb and output=http, in particular for metrics_written and buffer_size, just as when using telegraf 1.18.3. Gathered metric points should be buffered for both unavailable outputs, but only http output is buffered if using 1.19.1.

### Actual behavior:
When using 1.19.1 binary internal_write reports high number for metrics_written even though influxdb is not up and low number for buffer_size as it wasn't buffering the gathered points.
internal_write,host=xxx,output=influxdb,...,version=1.19.1 metrics_written=29000i,...,buffer_size=682i

On the contrary, 1.18.3 binary works as expected.
internal_write,host=xxxxx,output=influxdb,...,version=1.18.3 ...buffer_size=29718i,...,metrics_written=0i

http output also works as expected for both binary versions:
internal_write,host=xxxxx,output=http,...,version=1.19.1 ...buffer_size=29682i,...,metrics_written=0i
internal_write,host=xxxxx,output=http,...,version=1.18.3 ...,metrics_written=0i,...,buffer_size=29718i

### PR DESCRIPTION
**Pull request by MyaLongmire**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

resolves #9514 

Reworked code to handle more than database not found errors so it will not say metrics are getting pushed when the database goes down. Also updated test

### CODE REVIEW

File: plugins/outputs/influxdb/http_test.go [hunk 1]
```diff
@@ -1087,6 +1101,7 @@ func TestDBRPTagsCreateDatabaseCalledOnDatabaseNotFound(t *testing.T) {
 				}
 			},
 			func(w http.ResponseWriter, r *http.Request) {
+				fmt.Println("two:", r.URL.Path, r.FormValue("q"))
```
Comment (srebhan) [inline plugins/outputs/influxdb/http_test.go:1104]: I guess we want to get rid of those debugging left-overs. ;-)

File: plugins/outputs/influxdb/influxdb.go [hunk 2]
```diff
@@ -224,17 +224,19 @@ func (i *InfluxDB) Write(metrics []telegraf.Metric) error {
...
+				continue
 			}
+
+			err := client.CreateDatabase(ctx, apiError.Database)
+			if err != nil {
+				i.Log.Errorf("When writing to [%s]: database %q not found and failed to recreate",
+					client.URL(), apiError.Database)
+			} else {
+				return errors.New("database created; retry write")
+			}
```
Comment (srebhan) [inline plugins/outputs/influxdb/influxdb.go:237]: Why do you return an error on success but only log on failure? This seems strange here, so I would at least expect a solid comment explaining the _why_.

File: plugins/outputs/influxdb/influxdb.go [hunk 3]
```diff
@@ -295,6 +297,7 @@ func (i *InfluxDB) httpClient(ctx context.Context, url *url.URL, proxy *url.URL)
 	}
 
 	if !i.SkipDatabaseCreation {
+		fmt.Println("I am the request in http client")
```
Comment (srebhan) [inline plugins/outputs/influxdb/influxdb.go:300]: :-)

File: plugins/outputs/influxdb/influxdb.go [hunk 4]
```diff
@@ -224,17 +224,19 @@ func (i *InfluxDB) Write(metrics []telegraf.Metric) error {
...
-				err := client.CreateDatabase(ctx, apiError.Database)
-				if err != nil {
-					i.Log.Errorf("When writing to [%s]: database %q not found and failed to recreate",
-						client.URL(), apiError.Database)
-				} else {
-					// try another client, if all clients fail with this error, do not return error
-					continue
-				}
+			if i.SkipDatabaseCreation {
+				continue
```
Comment (srebhan) [inline plugins/outputs/influxdb/influxdb.go:228]: Can we please at least log a warning here that we dropped all those nice metrics?

**Review by srebhan — COMMENTED** [review]
Hey @MyaLongmire, thanks for the PR! There are some debug messages left in the code and I'd like to ask you to add a comment on the "strange" error handling. I guess it has to be like this for telegraf to trigger a retry, however it might be nicer to retry directly after successfully creating the database.
Assume your metric buffer is full and that's why a flush/write was triggered in the first place. In this first try you create the database and return an error (as you currently do). Now your metric buffer is still full. If new metrics arrive between this and the next retry you will loose metrics due to buffer fullness...

File: plugins/outputs/influxdb/influxdb.go (hunk 4 above)
Comment (MyaLongmire) [reply plugins/outputs/influxdb/influxdb.go:228]: This is just to remove the awful double negative that was `!i.SkipDatabaseCreation`.

File: plugins/outputs/influxdb/http_test.go (hunk 1 above)
Comment (MyaLongmire) [reply plugins/outputs/influxdb/http_test.go]: removed

File: plugins/outputs/influxdb/influxdb.go (hunk 2 above)
Comment (MyaLongmire) [reply plugins/outputs/influxdb/influxdb.go:239]: Added a comment

**Comment by srebhan** [comment]
@MyaLongmire can you comment on the potential buffer fullness issue I mentioned. It's ok for me if we do not tackle it in this PR, but if this is not just hypothetical, we need to take care of it.

**Comment by MyaLongmire** [comment]
> @MyaLongmire can you comment on the potential buffer fullness issue I mentioned. It's ok for me if we do not tackle it in this PR, but if this is not just hypothetical, we need to take care of it.

@srebhan in my understanding if the buffer is full when new metrics come in the oldest are dropped, there nothing much to be done about that.

**Comment by srebhan** [comment]
@MyaLongmire imagine the case where the database does not yet exist on the InfluxDB side. Further assume the buffer runs full (e.g. due to a delayed start of the database service or whatever). In the ideal case, the output plugin would create the database and flush the metric to this newly created database. That's what I would expect as user.
However, from my understanding what the plugin does is, it creates  the database in a first flush cycle **without** flushing the data to the newly created database and waits for a **second** flush cycle to write the data, The "data" in the second cycle comprises the "old" one from the first cycle and the new data arrived in the meantime. If now the buffer is already full (or almost full) in the first cycle, the data arriving between the first and second cycle might lead to data-dropping.

My suggestion is to change the current state
```golang
plugin.Write() // --> will create the database and return error without flushing the buffer
... gathering of data...
plugin.Write() // --> will flush the buffer containing data of the first "Write()" call and new data
```
to 
```golang
plugin.Write() // --> will create the database and try flushing the buffer
... gathering of data...
plugin.Write() // --> will flush the buffer containing ONLY new data
```

**Review by ssoroka — APPROVED** [review]

**Comment by ssoroka** [comment]
I get what you're saying @srebhan , but I don't think it's very likely that the buffer is full and the database has not yet been created but will be created successfully on this call. If that is the case, it's likely the buffer was already full and dropping messages.
I'm a bit hesitant to repeat the write immediately as the code is not very straight-forward. the choice to return an error to trigger retry is one chosen to keep the code simple in an area we've already had a lot of issues with code complexity.

**Comment by srebhan** [comment]
@ssoroka understood. It's fine with me.

**Review by srebhan — APPROVED** [review]
Looks good to me.

**Pull request merged**